lib/IRGen/IRGenFunction.cpp

```diff
@@ -513,3 +514,253 @@ llvm::Value *IRGenFunction::alignUpToMaximumAlignment(llvm::Type *sizeTy, llvm::
   auto *invertedMask = Builder.CreateNot(alignMask);
   return Builder.CreateAnd(Builder.CreateAdd(val, alignMask), invertedMask);
 }
+
+/// Returns the current task \p currTask as an UnsafeContinuation at +1.
+static llvm::Value *unsafeContinuationFromTask(IRGenFunction &IGF,
+                                               SILType unsafeContinuationTy,
+                                               llvm::Value *currTask) {
+  auto &IGM = IGF.IGM;
+  auto &Builder = IGF.Builder;
+
+  auto &rawPonterTI = IGM.getRawPointerTypeInfo();
+  auto object =
+      Builder.CreateBitOrPointerCast(currTask, rawPonterTI.getStorageType());
+
+  // Wrap the native object in the UnsafeContinuation struct.
+  //   struct UnsafeContinuation<T> {
+  //     let _continuation : Builtin.RawPointer
+  //   }
+  auto &unsafeContinuationTI =
+      cast<LoadableTypeInfo>(IGF.getTypeInfo(unsafeContinuationTy));
+  auto unsafeContinuationStructTy =
+      cast<llvm::StructType>(unsafeContinuationTI.getStorageType());
+  auto fieldTy =
+      cast<llvm::StructType>(unsafeContinuationStructTy->getElementType(0));
+  auto reference =
+      Builder.CreateBitOrPointerCast(object, fieldTy->getElementType(0));
+  auto field =
+      Builder.CreateInsertValue(llvm::UndefValue::get(fieldTy), reference, 0);
+  auto unsafeContinuation = Builder.CreateInsertValue(
+      llvm::UndefValue::get(unsafeContinuationStructTy), field, 0);
+
+  return unsafeContinuation;
+}
+
+void IRGenFunction::emitGetAsyncContinuation(SILType unsafeContinuationTy,
+                                             StackAddress resultAddr,
+                                             Explosion &out) {
+  // Create the continuation.
+  // void current_sil_function(AsyncTask *currTask, Executor *currExecutor,
+  //                           AsyncContext *currCtxt) {
+  //
+  // A continuation is the current AsyncTask 'currTask' with:
+  //   currTask->ResumeTask = @llvm.coro.async.resume();
+  //   currTask->ResumeContext = &continuation_context;
+  //
+  // Where:
+  //
+  // struct {
+  //   AsyncContext *resumeCtxt;
+  //   void *awaitSynchronization;
+  //   SwiftError *errResult;
+  //   union {
+  //     IndirectResult *result;
+  //     DirectResult *result;
```

Comment:
What's the difference between IndirectResult and DirectResult?

Reply:
Nothing, really.

Sometimes, we store direct results directly in a data structure.

In such a case I would have written:
```
struct {
   //   AsyncContext *resumeCtxt;
   //   void *awaitSynchronization;
   //   SwiftError *errResult;
   //   union {
   //     IndirectResult *result;
   //     DirectResult result;
```

I wanted to point out that the storage for both cases is the same.

Reply:
I see, so for a word-sized result, we store it directly in the context? That makes sense.

Reply:
No we don't currently.

With the sentence:

>Sometimes, we store direct results directly in a data structure.

I wanted to say: In other places such as the context passed to async functions (not the continuation context mentioned here) we store direct results in place and indirect results by address:
https://github.com/apple/swift/blob/main/lib/IRGen/GenCall.cpp#L161
https://github.com/apple/swift/blob/main/lib/IRGen/GenCall.cpp#L190

The notation:
```
  //   union {
  //     IndirectResult *result;
  //     DirectResult *result;
```
is meant to convey that both direct and indirect results are stored the same (as addresses to memory that stores the result).

Reply:
On net, I think just saying `Result*` would probably be clearer.

Reply:
https://github.com/apple/swift/pull/34937